Userland/Libraries/LibWeb/Fetch/Infrastructure/HTTP/Responses.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2022, Linus Groh <linusg@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibWeb/Fetch/Infrastructure/HTTP/Responses.h>
+
+namespace Web::Fetch {
+
+// https://fetch.spec.whatwg.org/#ref-for-concept-network-error%E2%91%A3
+// A network error is a response whose status is always 0, status message is always
+// the empty byte sequence, header list is always empty, and body is always null.
+
+Response Response::aborted_network_error()
+{
+    auto response = network_error();
+    response.set_aborted(true);
+    return response;
+}
+
+Response Response::network_error()
+{
+    Response response;
+    response.set_status(0);
+    response.set_type(Type::Error);
+    VERIFY(!response.body().has_value());
+    return response;
+}
+
+// https://fetch.spec.whatwg.org/#concept-aborted-network-error
+bool Response::is_aborted_network_error() const
+{
+    // A response whose type is "error" and aborted flag is set is known as an aborted network error.
+    return m_type == Type::Error && m_aborted;
+}
+
+// https://fetch.spec.whatwg.org/#concept-network-error
+bool Response::is_network_error() const
+{
+    // A response whose type is "error" is known as a network error.
+    return m_type == Type::Error;
+}
+
+// https://fetch.spec.whatwg.org/#concept-response-url
+Optional<AK::URL const&> Response::url() const
+{
+    // A response has an associated URL. It is a pointer to the last URL in response’s URL list and null if response’s URL list is empty.
+    if (m_url_list.is_empty())
+        return {};
+    return m_url_list.last();
+}
+
+// https://fetch.spec.whatwg.org/#concept-response-location-url
+ErrorOr<Optional<AK::URL>> Response::location_url(Optional<String> const& request_fragment) const
+{
+    // The location URL of a response response, given null or an ASCII string requestFragment, is the value returned by the following steps. They return null, failure, or a URL.
+
+    // 1. If response’s status is not a redirect status, then return null.
+    if (!is_redirect_status(m_status))
+        return Optional<AK::URL> {};
+
+    // FIXME: 2. Let location be the result of extracting header list values given `Location` and response’s header list.
+    auto location_value = ByteBuffer {};
+
+    // 3. If location is a header value, then set location to the result of parsing location with response’s URL.
+    auto location = AK::URL { StringView { location_value } };
+    if (location.is_valid())
```

Comment:
Is this backwards?
```suggestion
    if (!location.is_valid())
```

Reply:
:catdogalert:
Fixed, thanks!

Reply:
(it goes without saying that I haven't actually tested most of the fetch code yet - hopefully starting to add some JS APIs soon to make that easier)